File: graficacion/01/java/Circunferencia.java
public class Circunferencia extends Grafico {
  private Punto pCentral;
  private int radio;

  public Circunferencia(Punto p, int r, Panel canvas) {
    this.pCentral = p;
    this.radio = r;
    this.canvas = canvas;
  }

  // PUNTO-MEDIO
  @Override
  public void dibujarContorno(int grosor, Estilo estilo, XColor color, int delay) {
    int x = this.radio,
        y = 0,
        xChange = 1 - (this.radio << 1),
        yChange = 1,
        radiusError = 0;
    
    int pintar = 0, noPintar = 0, cnt = 1;
    if (estilo != null) {
      pintar = estilo.pintar;
      noPintar = estilo.noPintar;
    }

    boolean f = true;
    while (x >= y) {
      if (estilo != null) {
        if (cnt == 1) {
          pintar--;

          if (pintar == 0) {
            pintar = estilo.pintar;
            cnt = 0;
            f = false;
          }
        } else {
          noPintar--;

          if (noPintar == 0) {
            noPintar = estilo.noPintar;
            cnt = 1;
            f = true;
          }
        }
      } else
        f = true;


      if (f)
        addPoints(x, y, grosor, delay, color.getValor());

      y++;
      radiusError += yChange;
      yChange += 2;
      if ((radiusError << 1) + xChange > 0) {
        x--;
        radiusError += xChange;
        xChange += 2;
      }
    }
  }

  @Override
  public void dibujarRelleno(XColor color, int delay) { }


  private void addPoints(int x, int y, int grosor, int delay, int color) {
    for (int i = 0; i < grosor; i++)
      for (int j = 0; j < grosor; j++) {
        canvas.drawPixel(pCentral.x + x + i, pCentral.y + y + j, color);
        canvas.drawPixel(pCentral.x - x + i, pCentral.y + y + j, color);
        canvas.drawPixel(pCentral.x + x + i, pCentral.y - y + j, color);
        canvas.drawPixel(pCentral.x - x + i, pCentral.y - y + j, color);
        canvas.drawPixel(pCentral.x + y + i, pCentral.y + x + j, color);
        canvas.drawPixel(pCentral.x - y + i, pCentral.y + x + j, color);
        canvas.drawPixel(pCentral.x + y + i, pCentral.y - x + j, color);
        canvas.drawPixel(pCentral.x - y + i, pCentral.y - x + j, color);

        try {
          Thread.sleep(delay);
        } catch (InterruptedException e) {
          e.printStackTrace();
        }
      }
  }
}

File: graficacion/01/python/circunferencia.py
from canvas import Canvas
import time; delay = .01


class Circunferencia:
  def __init__(self, punto_c: tuple, radio: int) -> None:
    self.x_c, self.y_c = punto_c
    self.r = radio


  def midpoint(self, l: Canvas, color: str = '#fff', delay=.01) -> list:
    out = []
    x = self.r
    y = 0
    P = 1 - self.r

    # time.sleep(delay)
    # l.draw_pixel(x + self.x_c, y + self.y_c, c=color)
    # out.append((x + self.x_c, y + self.y_c))

    while x > y:
      y += 1

      if P <= 0:
        P = P + (y << 1) + 1
      else:
        x -= 1
        P = P + (y << 1) - (x << 1) + 1

      if x < y:
        break

      tmp = [
        (self.x_c + x, self.y_c + y), (self.x_c - x, self.y_c + y),
        (self.x_c + x, self.y_c - y), (self.x_c - x, self.y_c - y),
      ]

      if x != y:
        tmp.extend([
          (self.x_c + y, self.y_c + x), (self.x_c - y, self.y_c + x),
          (self.x_c + y, self.y_c - x), (self.x_c - y, self.y_c - x),
        ])

      for x_, y_ in tmp:
        time.sleep(delay)
        l.draw_pixel(x_, y_, c=color)
        out.append((x_, y_))

    return out


  def bresenham(self, l: Canvas, color: str = '#fff', delay=.01) -> list:
    out = []

    def draw(x, y) -> None:
      tmp = [
        (self.x_c + x, self.y_c + y), (self.x_c - x, self.y_c + y),
        (self.x_c + x, self.y_c - y), (self.x_c - x, self.y_c - y),
        (self.x_c + y, self.y_c + x), (self.x_c - y, self.y_c + x),
        (self.x_c + y, self.y_c - x), (self.x_c - y, self.y_c - x),
      ]

      for x_, y_ in tmp:
        time.sleep(delay)
        l.draw_pixel(x_, y_, c=color)
        out.append((x_, y_))

    x = 0
    y = self.r
    d = 3 - (self.r << 1)
    draw(x, y)

    while y >= x:
      x += 1

      if d > 0:
        y -= 1
        d = d + ((x - y) << 2) + 10
      else:
        d = d + (x << 2) + 6

      draw(x, y)

    return out
